### PR TITLE
Mr zephyr speech

### PR DESCRIPTION
**Pull request by szymon-filipiak**
Now hopefully this branch is without doubled commits. We need to resolve some conflicts regarding Makefile, CMake, project.conf and audio_provider.cc

### CODE REVIEW

File: tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+RecognizeCommands::RecognizeCommands(tflite::ErrorReporter* error_reporter,
+                                     int32_t average_window_duration_ms,
+                                     uint8_t detection_threshold,
+                                     int32_t suppression_ms,
+                                     int32_t minimum_count)
+    : error_reporter_(error_reporter),
+      average_window_duration_ms_(average_window_duration_ms),
+      detection_threshold_(detection_threshold),
+      suppression_ms_(suppression_ms),
+      minimum_count_(2),
```
Comment (plonajakub) [inline tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc:29]: Line 29 was originally the only change in this file. I suggest tweaking these parameters in the header (recognize_commands.h) and preserving the initialization list with the constructor's parameters.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,116 @@
+cmake_minimum_required(VERSION 3.13.1)
```
Comment (plonajakub) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/CMakeLists.txt]: I don't think we need to merge this CMake. We should rather stick to the convention and move changed TF files to the src directory inside the Zephyr's project directory.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/Makefile.inc [hunk 3]
```diff
@@ -0,0 +1,8 @@
+ifeq ($(TARGET), zephyr_vexriscv)
```
Comment (plonajakub) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/Makefile.inc]: I've customized the build process so this file has to be defined (and it is actually) a bit different. I think we should not merge this file.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/board/litex_vexriscv.overlay [hunk 4]
```diff
@@ -0,0 +1,4 @@
+&eth0 {
```
Comment (plonajakub) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/board/litex_vexriscv.overlay]: I see that we have this change in our project. Must be some git history quirk. Don't merge it.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/prj.conf [hunk 5]
```diff
@@ -0,0 +1,5 @@
+CONFIG_CPLUSPLUS=y
+CONFIG_NEWLIB_LIBC=y
+CONFIG_LOG=n
```
Comment (plonajakub) [inline tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/prj.conf:3]: This file adds `CONFIG_LOG=n` and `CONFIG_I2S_LITEX_DATA_BIG_ENDIAN=y`. The other lines are already in our project. I suspect that I2S configuration is necessary but I am not sure what `CONFIG_LOG` does.

File: tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc (hunk 1 above)
Comment (pawelsag) [reply tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc:29]: Yeap, it for sure should be changed in the header file where the default parameters are set.
But this is not the solution, we should improve detection, cause the default value is set to 3. In my opinion, we shouldn't lower this value and look for a solution that will improve detection. What I try to say is that the value is low enough and tweaking default values won't bring expected results. (probably)
Finally, I have some time to try to figure out what operations may improve results.
Do you have any idea? In your opinion, how accurate the net is and what was the results for "no" word, when you played with it.

File: tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/board/litex_vexriscv.overlay (hunk 4 above)
Comment (pawelsag) [reply tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/board/litex_vexriscv.overlay]: I guess you shouldn't merge this file at all, this file is specific to your device configuration

File: tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc (hunk 1 above)
Comment (plonajakub) [reply tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc:29]: I am happy that the place for changes like this is clear now :). My intention here was only to point this out. Anyway, I agree that we should investigate the topic. Summon @zawadzki-dawid for results of the training. @zawadzki-dawid, have you seen or can share some final reports about the net's accuracy from the training script?

**Review by pawelsag — COMMENTED** [review]
CONFIG_LOG turns off logs generated by functions usually used in drivers
CONFIG_I2S_LITEX_DATA_BIG_ENDIAN enables putting data received by the driver in big-endian. Such layout is required by TF core @plonajakub 
I don't know why my comments are pending when I tried to answer in your comments

File: tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc (hunk 1 above)
Comment (pawelsag) [reply tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc:29]: I thought about our problem and I'm pretty sure we are receiving data too fast, I know how to increase the value of the `minimum_count_`  variable. Moreover, I think we shouldn't touch any values in class definition/declaration, other parts of code can use it and we will break the compatibility, we should pass the desired variables during object construction in the object which we are using.

Comment (szymon-filipiak) [reply tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc:29]: One could argue that changing these values only in constructed objects can lead to a mismatch across different classes. I opt for changing them in the header files. As for the receiving data too fast: I've spoken with Jakub and he has said that we've had a problem with time measured by TF but we managed to resolve it. @plonajakub Can you confirm?

Comment (zawadzki-dawid) [reply tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc:29]: @plonajakub i have wrote it down somewhere at the Desktop but I lost it somehow :/. But as far as I remember test after training had 74.3% of accuracy.

**Review by zawadzki-dawid — APPROVED** [review]

File: tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc (hunk 1 above)
Comment (plonajakub) [reply tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc:29]: > One could argue that changing these values only in constructed objects can lead to a mismatch across different classes. I opt for changing them in the header files. As for the receiving data too fast: I've spoken with Jakub and he has said that we've had a problem with time measured by TF but we managed to resolve it. @plonajakub Can you confirm?

For the question: there is a high probability that we are (with current implementation in voice-assistant) on the right track. The corresponding change is here: 88ff3957bedde8db0f3fcf3856bf1a39e26340c8. Note, that this was not tested on a real device (but was tested in Renode). Anyway, I doubt that the timing will solve the problem with bad accuracy and I am not sure if it is related to the problem with receiving the data too fast.

>@plonajakub i have wrote it down somewhere at the Desktop but I lost it somehow :/. But as far as I remember test after training had 74.3% of accuracy.

Also, 74% is not really that high when it comes to the accuracy of a classifier. I think we should discuss it in a meeting.

**Comment by plonajakub** [comment]
> I don't know why my comments are pending when I tried to answer in your comments

I have no idea also :D. I haven't changed any posting rules.
Thanks for the explanation. Consider it resolved.

**Comment by plonajakub** [comment]
Be informed that a newer version of tensorflow/lite/micro/examples/micro_speech/zephyr_riscv/src/audio_provider.cc is already on voice-assistant. The tensorflow/lite/micro/examples/micro_speech/recognize_commands.h replaces tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc (the change is also on voice-assistant already).

The only change left here to merge is prj.conf.

Reference: 75aba80edac2550ac8e58aa39cd8ed256e07ec51.

**Review by Marzecki — APPROVED** [review]
